{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "334a81e4-d695-458c-83fc-07784dcdc821",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import re\n",
    "import seaborn as sns\n",
    "from tqdm import tqdm\n",
    "from statannotations.Annotator import Annotator\n",
    "from scipy.stats import mannwhitneyu\n",
    "from plottable import ColumnDefinition, Table\n",
    "from plottable.plots import bar\n",
    "from plottable.cmap import normed_cmap, centered_cmap\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib_venn import venn2, venn2_circles\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import gseapy as gp\n",
    "from plottable import ColumnDefinition, Table\n",
    "from plottable.cmap import normed_cmap\n",
    "from plottable.formatters import decimal_to_percent, tickcross\n",
    "from plottable.plots import circled_image\n",
    "import matplotlib\n",
    "import matplotlib.colors as mcolors\n",
    "import pathlib"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44d75961",
   "metadata": {},
   "source": [
    "# Setup path and params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddbf03f4-19d4-4d39-8a55-457e7e62c527",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_genetics_1 = \"E:/YandexDisk/Work/bbd/fmba/exome\"\n",
    "path_genetics_2 = \"E:/YandexDisk/Work/bbd/fmba/lesnoy_2025/results/tapes_annotations_2025\"\n",
    "path_main = 'E:/YandexDisk/Work/pydnameth/draft/13_fmba_cvd_dnam/data/120_1'\n",
    "\n",
    "sheet_name_1 = \"top200\"\n",
    "sheet_name_2 = \"Top 300\"\n",
    "prob_thld = 0.8\n",
    "\n",
    "path_save = f\"{path_main}/genetics/{prob_thld}\"\n",
    "pathlib.Path(f\"{path_save}/gsea_libs\").mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7243f4c1",
   "metadata": {},
   "source": [
    "# Generate and save mutations dataframe from first batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4669e8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gen_1 = pd.read_excel(f\"{path_genetics_1}/Образцы экзомы Лесной секвенирование 100 шт.xlsx\", index_col='Sample#')\n",
    "df_gen_1['ID образца крови'] = df_gen_1['ID образца крови'].astype(str)\n",
    "\n",
    "\n",
    "all_results = []\n",
    "for file in (pbar := tqdm(os.listdir(path_genetics_1))):\n",
    "    pbar.set_description(f\"{file}\")\n",
    "    \n",
    "    if file.endswith(\".xlsx\"):\n",
    "        file_path = os.path.join(path_genetics_1, file)\n",
    "\n",
    "        try:\n",
    "            df = pd.read_excel(file_path, sheet_name=sheet_name_1)\n",
    "            \n",
    "            # Убедимся, что нужные колонки присутствуют\n",
    "            required_cols = [\n",
    "                \"Ген\", \"Вероятность патогенности\", \"ACMG_класс\", \"Ассоциация с заболеванием\"\n",
    "            ]\n",
    "            if not all(col in df.columns for col in required_cols):\n",
    "                print(f\"Пропущен файл {file_path} — нет нужных колонок.\")\n",
    "                continue\n",
    "            \n",
    "            # Фильтрация\n",
    "            filtered = df[df[\"Вероятность патогенности\"] >= prob_thld].copy()\n",
    "            \n",
    "            sample_number_raw = file.split(\"-\")[0]\n",
    "            sample_number = df_gen_1.at[sample_number_raw, 'ID образца крови']\n",
    "            if not filtered.empty:\n",
    "                \n",
    "                filtered[\"Номер образца\"] = sample_number\n",
    "\n",
    "                all_results.append(filtered[required_cols + [\"Номер образца\"]])\n",
    "            else:\n",
    "                print(sample_number)\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"Ошибка при обработке файла {file_path}: {e}\")\n",
    "\n",
    "# Объединяем в один датафрейм\n",
    "if all_results:\n",
    "    mutations_df = pd.concat(all_results, ignore_index=True)\n",
    "\n",
    "    # Переименование колонок\n",
    "    mutations_df.rename(columns={\n",
    "        \"Ген\": \"Gene\",\n",
    "        \"Вероятность патогенности\": \"Pathogenicity Probability\",\n",
    "        \"ACMG_класс\": \"ACMG Classification\",\n",
    "        \"Ассоциация с заболеванием\": \"Disease Description\",\n",
    "        \"Номер образца\": \"Sample\"\n",
    "    }, inplace=True)\n",
    "else:\n",
    "    print(\"Подходящих данных не найдено.\")\n",
    "\n",
    "mutations_df[\"number\"] = mutations_df[\"Sample\"].astype(str)\n",
    "mutations_df.to_excel(f\"{path_save}/filtered_mutations_1.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "374f5230",
   "metadata": {},
   "source": [
    "# Generate and save mutations dataframe from second batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff3d69b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_results = []\n",
    "for file in (pbar := tqdm(os.listdir(path_genetics_2))):\n",
    "    pbar.set_description(f\"{file}\")\n",
    "    \n",
    "    if file.endswith(\".vcf.xlsx\"):\n",
    "        file_path = os.path.join(path_genetics_2, file)\n",
    "\n",
    "        try:\n",
    "            df = pd.read_excel(file_path, sheet_name=sheet_name_2)\n",
    "            \n",
    "            # Убедимся, что нужные колонки присутствуют\n",
    "            required_cols = [\n",
    "                \"Ген\", \"Вероятность патогенности\", \"Классификация ACMG\", \"Описание заболевания\"\n",
    "            ]\n",
    "            if not all(col in df.columns for col in required_cols):\n",
    "                print(f\"Пропущен файл {file_path} — нет нужных колонок.\")\n",
    "                continue\n",
    "            \n",
    "            # Фильтрация\n",
    "            filtered = df[df[\"Вероятность патогенности\"] >= prob_thld].copy()\n",
    "            \n",
    "            sample_number = file.split(\".\")[0]\n",
    "            if not filtered.empty:\n",
    "                \n",
    "                filtered[\"Номер образца\"] = sample_number\n",
    "\n",
    "                all_results.append(filtered[required_cols + [\"Номер образца\"]])\n",
    "            else:\n",
    "                print(sample_number)\n",
    "\n",
    "        except Exception as e:\n",
    "            print(f\"Ошибка при обработке файла {file_path}: {e}\")\n",
    "\n",
    "# Объединяем в один датафрейм\n",
    "if all_results:\n",
    "    mutations_df = pd.concat(all_results, ignore_index=True)\n",
    "\n",
    "    # Переименование колонок\n",
    "    mutations_df.rename(columns={\n",
    "        \"Ген\": \"Gene\",\n",
    "        \"Вероятность патогенности\": \"Pathogenicity Probability\",\n",
    "        \"Классификация ACMG\": \"ACMG Classification\",\n",
    "        \"Описание заболевания\": \"Disease Description\",\n",
    "        \"Номер образца\": \"Sample\"\n",
    "    }, inplace=True)\n",
    "else:\n",
    "    print(\"Подходящих данных не найдено.\")\n",
    "\n",
    "mutations_df[\"number\"] = mutations_df[\"Sample\"].astype(str)\n",
    "mutations_df.to_excel(f\"{path_save}/filtered_mutations_2.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1da0a403",
   "metadata": {},
   "source": [
    "# Load generated mutations dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9688698e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_df_1 = pd.read_excel(f\"{path_save}/filtered_mutations_1.xlsx\")\n",
    "mutations_df_2 = pd.read_excel(f\"{path_save}/filtered_mutations_2.xlsx\")\n",
    "samples_1 = mutations_df_1[\"Sample\"].unique()\n",
    "samples_2 = mutations_df_2[\"Sample\"].unique()\n",
    "print(set(samples_1) & set(samples_2))\n",
    "df_samples = pd.DataFrame(index=list(set(samples_1) | set(samples_2)))\n",
    "df_samples.loc[samples_1, 'Batch'] = 1\n",
    "df_samples.loc[samples_2, 'Batch'] = 2\n",
    "df_samples.to_excel(f\"{path_save}/df_samples.xlsx\")\n",
    "\n",
    "mutations_df = pd.concat([mutations_df_1, mutations_df_2], ignore_index=True)\n",
    "\n",
    "mutations_df[\"number\"] = mutations_df[\"Sample\"].astype(str)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ec3720b",
   "metadata": {},
   "source": [
    "# Merge genetics and pheno dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddd8a041",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_groups_df = pd.read_excel(f\"{path_main}/pheno_funnorm.xlsx\", index_col=0)\n",
    "patient_groups_df[\"number\"] = patient_groups_df.index.astype(str)\n",
    "patient_groups_df[\"Status\"] = patient_groups_df[\"Special Status\"]\n",
    "display(patient_groups_df[\"Status\"].value_counts())\n",
    "\n",
    "samples = list(set(mutations_df['number'].unique()).intersection(set(patient_groups_df['number'].unique())))\n",
    "patient_groups_df = patient_groups_df[patient_groups_df[\"number\"].isin(samples)]\n",
    "display(patient_groups_df[\"Status\"].value_counts())\n",
    "\n",
    "# Объединяем по колонке number\n",
    "merged_df = mutations_df.merge(patient_groups_df, on=\"number\")\n",
    "\n",
    "# Среднее число мутаций на пациента в каждой группе\n",
    "display(merged_df.groupby(\"Status\")[\"Gene\"].count() / merged_df.groupby(\"Status\")[\"Sample\"].nunique())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78809b15",
   "metadata": {},
   "source": [
    "# Average mutations count in groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecea454f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_grouped = merged_df.groupby([\"Status\", \"Sample\"])[\"Gene\"].count().reset_index(name=\"Mutation count\")\n",
    "\n",
    "sns.set_theme(style='ticks')\n",
    "fig, ax = plt.subplots(figsize=(4, 4), layout=\"constrained\")\n",
    "violin = sns.violinplot(\n",
    "    data=df_grouped,\n",
    "    x='Status',\n",
    "    y='Mutation count',\n",
    "    palette={'Control': 'chartreuse', 'Case': 'red'},\n",
    "    scale='width',\n",
    "    order=['Control', 'Case'],\n",
    "    saturation=0.75,\n",
    ")\n",
    "violin.set_xlabel(f\"\")\n",
    "violin.set_ylabel(\"Average number of mutations per sample\")\n",
    "mw_pval = mannwhitneyu(\n",
    "    df_grouped.loc[df_grouped['Status'] == 'Control', f'Mutation count'].values,\n",
    "    df_grouped.loc[df_grouped['Status'] == 'Case', f'Mutation count'].values,\n",
    "    alternative='two-sided').pvalue\n",
    "pval_formatted = [f'{mw_pval:.2e}']\n",
    "annotator = Annotator(\n",
    "    violin,\n",
    "    pairs=[('Control', 'Case')],\n",
    "    data=df_grouped,\n",
    "    x='Status',\n",
    "    y=f'Mutation count',\n",
    "    order=['Control', 'Case']\n",
    ")\n",
    "annotator.set_custom_annotations(pval_formatted)\n",
    "annotator.configure(loc='outside')\n",
    "annotator.annotate()\n",
    "fig.savefig(f\"{path_save}/violin_mutation_count.png\", bbox_inches='tight', dpi=200)\n",
    "fig.savefig(f\"{path_save}/violin_mutation_count.pdf\", bbox_inches='tight')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95d352d5",
   "metadata": {},
   "source": [
    "# Average mutations count in groups by ACMG Classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff01b5c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_grouped_acmg = merged_df.groupby([\"Status\", \"Sample\", \"ACMG Classification\"])[\"Gene\"].count().reset_index(name=\"Mutation count\")\n",
    "\n",
    "acmg_colors = {\n",
    "    'Likely Benign': 'dodgerblue',\n",
    "    'VUS': 'gold',\n",
    "    'Likely Pathogenic': 'orangered',\n",
    "    'Pathogenic': 'firebrick'\n",
    "}\n",
    "\n",
    "sns.set_theme(style='ticks')\n",
    "fig = plt.figure(\n",
    "    figsize=(6, 5),\n",
    "    layout=\"constrained\"\n",
    ")\n",
    "axs = fig.subplot_mosaic(\n",
    "    [\n",
    "        ['table'],\n",
    "        ['violin'],\n",
    "    ],\n",
    "    height_ratios=[1, 5],\n",
    "    # width_ratios=[3, 1.5],\n",
    "    gridspec_kw={\n",
    "        # \"bottom\": 0.0,\n",
    "        # \"top\": 1.00,\n",
    "        # \"left\": 0.1,\n",
    "        # \"right\": 0.5,\n",
    "        #\"wspace\": 0.33,\n",
    "        #\"hspace\": 0.01,\n",
    "    },\n",
    ")\n",
    "\n",
    "ds_table = pd.DataFrame(index=['Control VS Case'], columns=list(acmg_colors.keys()))\n",
    "for acmg_group in acmg_colors:\n",
    "    mw_pval = mannwhitneyu(\n",
    "        df_grouped_acmg.loc[(df_grouped_acmg['Status'] == 'Control') & (df_grouped_acmg['ACMG Classification'] == acmg_group), f'Mutation count'].values,\n",
    "        df_grouped_acmg.loc[(df_grouped_acmg['Status'] == 'Case') & (df_grouped_acmg['ACMG Classification'] == acmg_group), f'Mutation count'].values,\n",
    "        alternative='two-sided'\n",
    "    ).pvalue\n",
    "    ds_table.at['Control VS Case', acmg_group] = f'{mw_pval:.2e}'\n",
    "col_defs = [\n",
    "    ColumnDefinition(\n",
    "        name=\"index\",\n",
    "        title='Mann-Whitney p-value',\n",
    "        textprops={\"ha\": \"left\"},\n",
    "        width=4.5,\n",
    "    ),\n",
    "]\n",
    "for acmg_group in acmg_colors:\n",
    "    col_defs.append(\n",
    "            ColumnDefinition(\n",
    "            name=acmg_group,\n",
    "            title=acmg_group,\n",
    "            textprops={\"ha\": \"center\"},\n",
    "            width=2.0,\n",
    "        )\n",
    "    )\n",
    "table = Table(\n",
    "    ds_table,\n",
    "    column_definitions=col_defs,\n",
    "    row_dividers=True,\n",
    "    footer_divider=False,\n",
    "    ax=axs['table'],\n",
    "    textprops={\"fontsize\": 7},\n",
    "    row_divider_kw={\"linewidth\": 1, \"linestyle\": (0, (1, 1))},\n",
    "    col_label_divider_kw={\"linewidth\": 1, \"linestyle\": \"-\"},\n",
    "    column_border_kw={\"linewidth\": 1, \"linestyle\": \"-\"},\n",
    ")\n",
    "\n",
    "violin = sns.violinplot(\n",
    "    data=df_grouped_acmg,\n",
    "    x='Status',\n",
    "    y='Mutation count',\n",
    "    hue=\"ACMG Classification\",\n",
    "    palette=acmg_colors,\n",
    "    hue_order=list(acmg_colors.keys()),\n",
    "    order=['Control', 'Case'],\n",
    "    scale='width',\n",
    "    saturation=0.75,\n",
    "    ax=axs['violin']\n",
    ")\n",
    "axs['violin'].set_xlabel(f\"\")\n",
    "axs['violin'].set_ylabel(\"Average number of mutations per sample\")\n",
    "fig.savefig(f\"{path_save}/violin_mutation_acmg_count.png\", bbox_inches='tight', dpi=200)\n",
    "fig.savefig(f\"{path_save}/violin_mutation_acmg_count.pdf\", bbox_inches='tight')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4fb9e99",
   "metadata": {},
   "source": [
    "# Frequency of genomic variants in the 50 most frequently mutated genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04e45829",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Подсчитываем количество мутаций для каждого гена в обеих группах\n",
    "mutation_counts = merged_df.groupby([\"Status\", \"Gene\"]).size().unstack(fill_value=0)\n",
    "\n",
    "# Считаем общее количество мутаций для каждого гена (по обеим группам)\n",
    "total_mutation_counts = mutation_counts.sum(axis=0)\n",
    "\n",
    "# Выбираем топ-50 генов с наибольшим количеством мутаций\n",
    "top_50_genes = total_mutation_counts.nlargest(50).index\n",
    "\n",
    "# Отбираем данные для этих топ-50 генов\n",
    "mutation_counts_filtered = mutation_counts.loc[:, top_50_genes]\n",
    "\n",
    "\n",
    "df_heatmap = mutation_counts_filtered.T[['Control', 'Case']]\n",
    "\n",
    "sns.set_theme(style='ticks')\n",
    "fig, ax = plt.subplots(figsize=(4, 15) , layout='constrained')\n",
    "heatmap = sns.heatmap(\n",
    "    df_heatmap,\n",
    "    annot=True,\n",
    "    fmt=\"d\",\n",
    "    # cmap='coolwarm',\n",
    "    linewidth=0.1,\n",
    "    linecolor='black',\n",
    "    # annot_kws={\"fontsize\": 15},\n",
    "    cbar_kws={\n",
    "        'orientation': 'horizontal',\n",
    "        'location': 'top',\n",
    "        'pad': 0.025,\n",
    "        'aspect': 30\n",
    "    },\n",
    "    ax=ax\n",
    ")\n",
    "heatmap_pos = heatmap.get_position()\n",
    "ax.figure.axes[-1].set_title('Frequency of genomic variants in\\nthe 50 most frequently mutated genes', fontsize=10)\n",
    "ax.figure.axes[-1].tick_params()\n",
    "for spine in ax.figure.axes[-1].spines.values():\n",
    "    spine.set_linewidth(1)\n",
    "ax.set_xlabel('')\n",
    "ax.set_ylabel('')\n",
    "plt.savefig(f\"{path_save}/heatmap_top50.png\", bbox_inches='tight', dpi=200)\n",
    "plt.savefig(f\"{path_save}/heatmap_top50.pdf\", bbox_inches='tight')\n",
    "plt.close(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2f17002",
   "metadata": {},
   "source": [
    "# GSEA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69a7e33b",
   "metadata": {},
   "source": [
    "## Prepare gene lists and libs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "095f55b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Получаем множества генов для каждой группы\n",
    "genes_ctrl = set(merged_df[merged_df['Status'] == 'Control']['Gene'].dropna())\n",
    "genes_case = set(merged_df[merged_df['Status'] == 'Case']['Gene'].dropna())\n",
    "\n",
    "genes_ctrl_only = genes_ctrl - genes_case\n",
    "genes_case_only = genes_case - genes_ctrl\n",
    "genes_cmn = genes_ctrl & genes_case\n",
    "\n",
    "gsea_libs_trgt = [\n",
    "    'GO_Biological_Process_2025',\n",
    "    'GO_Molecular_Function_2025',\n",
    "    'GO_Cellular_Component_2025',\n",
    "    'KEGG_2021_Human',\n",
    "    'Reactome_2022',\n",
    "    'Reactome_Pathways_2024',\n",
    "    'Jensen_DISEASES',\n",
    "    'Jensen_DISEASES_Curated_2025',\n",
    "    'Jensen_DISEASES_Experimental_2025',\n",
    "]\n",
    "\n",
    "genes_set = {\n",
    "    'ctrl_only': genes_ctrl_only,\n",
    "    'case_only': genes_case_only,\n",
    "    'cmn': genes_cmn\n",
    "}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 6))\n",
    "venn = venn2(\n",
    "    subsets=(genes_ctrl, genes_case),\n",
    "    set_labels = ('Control', 'Case'),\n",
    "    set_colors=('chartreuse', 'red'),\n",
    "    alpha = 0.8\n",
    ")\n",
    "venn2_circles(\n",
    "    subsets=(genes_ctrl, genes_case),\n",
    ")\n",
    "plt.tight_layout()\n",
    "plt.savefig(f\"{path_save}/venn_genes_intersection.png\", bbox_inches='tight', dpi=400)\n",
    "plt.savefig(f\"{path_save}/venn_genes_intersection.pdf\", bbox_inches='tight', dpi=400)\n",
    "plt.clf()\n",
    "\n",
    "# Печать статистики\n",
    "print(f\"Гены только у Control: {len(genes_ctrl_only)}\")\n",
    "print(f\"Гены только у Case: {len(genes_case_only)}\")\n",
    "print(f\"Общие гены: {len(genes_cmn)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1510f376",
   "metadata": {},
   "source": [
    "## Download enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "988690ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene_set_name, gene_set in genes_set.items():\n",
    "    dfs_enrichr = []\n",
    "    for gsea_lib in (pbar := tqdm(gsea_libs_trgt)):\n",
    "        pbar.set_description(f\"Processing {gsea_lib}\")\n",
    "        enr = gp.enrichr(\n",
    "            gene_list=list(gene_set),\n",
    "            gene_sets=gsea_lib,\n",
    "            organism='Human',\n",
    "            outdir=None,\n",
    "            cutoff=1.00,\n",
    "            verbose=False,\n",
    "            no_plot=True\n",
    "        )\n",
    "        dfs_enrichr.append(enr.results)\n",
    "    df_enrichr = pd.concat(dfs_enrichr, ignore_index=True)\n",
    "    df_enrichr.to_excel(f\"{path_save}/enrichr_{gene_set_name}.xlsx\", index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "421efe00",
   "metadata": {},
   "source": [
    "## Read enrichment from file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba218cbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs_enr = {}\n",
    "for gene_set_name in genes_set:\n",
    "    dfs_enr[gene_set_name] = pd.read_excel(f\"{path_save}/enrichr_{gene_set_name}.xlsx\", index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "437716af",
   "metadata": {},
   "source": [
    "## Terms analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b906a51",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gsea_lib in (pbar := tqdm(gsea_libs_trgt)):\n",
    "    pbar.set_description(f\"Processing {gsea_lib}\")\n",
    "    \n",
    "    terms_ctrl_only = set(dfs_enr['ctrl_only'].loc[(dfs_enr['ctrl_only']['Adjusted P-value'] < 0.05) & (dfs_enr['ctrl_only']['Gene_set'] == gsea_lib), 'Term'])\n",
    "    terms_case_only = set(dfs_enr['case_only'].loc[(dfs_enr['case_only']['Adjusted P-value'] < 0.05) & (dfs_enr['case_only']['Gene_set'] == gsea_lib), 'Term'])\n",
    "    terms_cmn = set(dfs_enr['cmn'].loc[(dfs_enr['cmn']['Adjusted P-value'] < 0.05) & (dfs_enr['cmn']['Gene_set'] == gsea_lib), 'Term'])\n",
    "    \n",
    "    terms_all = list(terms_ctrl_only.union(terms_case_only, terms_cmn))\n",
    "    max_letters_in_terms = max([len(x) for x in terms_all])\n",
    "    \n",
    "    term_presence = {\n",
    "        'Control only': [1 if term in terms_ctrl_only else 0 for term in terms_all],\n",
    "        'Case only': [1 if term in terms_case_only else 0 for term in terms_all],\n",
    "        'Common': [1 if term in terms_cmn else 0 for term in terms_all]\n",
    "    }\n",
    "    \n",
    "    df_term_presence = pd.DataFrame(term_presence, index=terms_all).sort_index()\n",
    "    \n",
    "    f_cmap = sns.color_palette(\"Spectral_r\", as_cmap=True)\n",
    "    f_norm = mcolors.Normalize(vmin=0, vmax=1)\n",
    "    sm = plt.cm.ScalarMappable(cmap=f_cmap, norm=f_norm).to_rgba\n",
    "    \n",
    "    col_defs = [\n",
    "        ColumnDefinition(\n",
    "            name=\"index\",\n",
    "            title='Terms',\n",
    "            textprops={\"ha\": \"left\"},\n",
    "            width=3.0 + 0.08 * max_letters_in_terms,\n",
    "        ),\n",
    "        ColumnDefinition(\n",
    "            name='Control only',\n",
    "            title='Control only',\n",
    "            formatter=tickcross,\n",
    "            textprops={\"ha\": \"center\"},\n",
    "            group=gsea_lib,\n",
    "            cmap=sm,\n",
    "            width=1.7,\n",
    "        ),\n",
    "        ColumnDefinition(\n",
    "            name='Case only',\n",
    "            title='Case only',\n",
    "            formatter=tickcross,\n",
    "            textprops={\"ha\": \"center\"},\n",
    "            group=gsea_lib,\n",
    "            cmap=sm,\n",
    "            width=1.7,\n",
    "        ),\n",
    "        ColumnDefinition(\n",
    "            name='Common',\n",
    "            title='Common',\n",
    "            formatter=tickcross,\n",
    "            textprops={\"ha\": \"center\"},\n",
    "            group=gsea_lib,\n",
    "            cmap=sm,\n",
    "            width=1.7,\n",
    "        ),\n",
    "    ]\n",
    "    \n",
    "    fig, ax = plt.subplots(figsize=(6.0 + 0.08 * max_letters_in_terms, 4 + 0.2 * df_term_presence.shape[0]))\n",
    "    table = Table(\n",
    "        df_term_presence,\n",
    "        column_definitions=col_defs,\n",
    "        row_dividers=True,\n",
    "        footer_divider=False,\n",
    "        odd_row_color=\"#ffffff\",\n",
    "        even_row_color=\"#f0f0f0\",\n",
    "        ax=ax,\n",
    "        # textprops={\"fontsize\": 10},\n",
    "        row_divider_kw={\"linewidth\": 1, \"linestyle\": (0, (1, 1))},\n",
    "        col_label_divider_kw={\"linewidth\": 1, \"linestyle\": \"-\"},\n",
    "        column_border_kw={\"linewidth\": 1, \"linestyle\": \"-\"},\n",
    "    ).autoset_fontcolors(colnames=['Control only', 'Case only', 'Common'])\n",
    "    fig.savefig(f\"{path_save}/gsea_libs/{gsea_lib}.png\", bbox_inches='tight', dpi=200)\n",
    "    fig.savefig(f\"{path_save}/gsea_libs/{gsea_lib}.pdf\", bbox_inches='tight')\n",
    "    plt.close(fig)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py312",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
